File: code_translation/exp1_code_translation_security/translated_solutions_gpt-4o/Question1/Go/1012.go
package main

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"log"

	"golang.org/x/crypto/chacha20poly1305"
)

var key []byte

func init() {
	// Generate a random 32-byte key for ChaCha20-Poly1305
	key = make([]byte, chacha20poly1305.KeySize)
	_, err := rand.Read(key)
	if err != nil {
		log.Fatalf("Failed to generate key: %v", err)
	}
}

func encryptText(plainText string) (string, error) {
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return "", fmt.Errorf("failed to create AEAD: %v", err)
	}

	nonce := make([]byte, chacha20poly1305.NonceSize)
	_, err = rand.Read(nonce)
	if err != nil {
		return "", fmt.Errorf("failed to generate nonce: %v", err)
	}

	encrypted := aead.Seal(nil, nonce, []byte(plainText), nil)
	// Combine nonce and encrypted text, then encode in base64
	result := append(nonce, encrypted...)
	return base64.StdEncoding.EncodeToString(result), nil
}

func decryptText(encryptedText string) (string, error) {
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return "", fmt.Errorf("failed to create AEAD: %v", err)
	}

	data, err := base64.StdEncoding.DecodeString(encryptedText)
	if err != nil {
		return "", fmt.Errorf("failed to decode base64: %v", err)
	}

	if len(data) < chacha20poly1305.NonceSize {
		return "", fmt.Errorf("invalid encrypted text")
	}

	nonce, ciphertext := data[:chacha20poly1305.NonceSize], data[chacha20poly1305.NonceSize:]
	plainText, err := aead.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("failed to decrypt: %v", err)
	}

	return string(plainText), nil
}

func main() {
	plainText := "Hello, World!"

	// Encrypt the text
	encryptedText, err := encryptText(plainText)
	if err != nil {
		log.Fatalf("Encryption failed: %v", err)
	}
	fmt.Println("Encrypted Text:", encryptedText)

	// Decrypt the text
	decryptedText, err := decryptText(encryptedText)
	if err != nil {
		log.Fatalf("Decryption failed: %v", err)
	}
	fmt.Println("Decrypted Text:", decryptedText)
}

File: code_translation/exp1_code_translation_security/translated_solutions_gpt-4o/Question1/Rust/1012.rs
use aes_gcm::{Aes256Gcm, Key, Nonce}; // Or `aes-gcm-siv` for SIV mode
use aes_gcm::aead::{Aead, NewAead};
use rand::Rng;

fn generate_key() -> Vec<u8> {
    let mut key = vec![0u8; 32]; // 256-bit key
    rand::thread_rng().fill(&mut key[..]);
    key
}

fn encrypt_text(plain_text: &str, key: &[u8]) -> Result<String, Box<dyn std::error::Error>> {
    let cipher = Aes256Gcm::new(Key::from_slice(key));
    let nonce = rand::thread_rng().gen::<[u8; 12]>(); // 96-bit nonce
    let nonce = Nonce::from_slice(&nonce);

    let encrypted_text = cipher.encrypt(nonce, plain_text.as_bytes())?;
    let mut result = nonce.to_vec();
    result.extend_from_slice(&encrypted_text);

    Ok(base64::encode(result))
}

fn decrypt_text(encrypted_text: &str, key: &[u8]) -> Result<String, Box<dyn std::error::Error>> {
    let cipher = Aes256Gcm::new(Key::from_slice(key));
    let encrypted_data = base64::decode(encrypted_text)?;
    let (nonce, ciphertext) = encrypted_data.split_at(12); // Split nonce and ciphertext
    let nonce = Nonce::from_slice(nonce);

    let decrypted_text = cipher.decrypt(nonce, ciphertext)?;
    Ok(String::from_utf8(decrypted_text)?)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let key = generate_key();

    let plain_text = "Hello, world!";
    let encrypted_text = encrypt_text(plain_text, &key)?;
    println!("Encrypted: {}", encrypted_text);

    let decrypted_text = decrypt_text(&encrypted_text, &key)?;
    println!("Decrypted: {}", decrypted_text);

    Ok(())
}